### ISSUE TITLE
java.util.regex.Matcher.replaceAll inconsistent on Native and JVM

### ISSUE THREAD

**Issue by olafurpg** [scala-native/scala-native#1576]
While testing Scalafmt on Native I hit on a case where `replaceAll` treats the replacement strings differently on Native compared to the JVM.

Here's my best effort to create a minimal reproduction
```scala
val leadingPipeSpace = Pattern.compile("\n *\\|", Pattern.MULTILINE)
leadingPipeSpace.matcher("\n    |\"").replaceAll(s"\n  \\|")
```

I was able to work around the problem by removing `replaceAll` and use `appendReplacement` instead
```scala
      val spaces = " " * (indent + extraIndent)
      val out = new java.lang.StringBuffer
      val replacement = s"\n$spaces|"
      val matcher = leadingPipeSpace.matcher(token.syntax)
      while (matcher.find()) {
        matcher.appendReplacement(out, replacement)
      }
      matcher.appendTail(out)
      out.toString()
```

**Comment by densh** [comment]
@LeeTibbert  Can you look at this one? Two main questions here:
1. Can we hotfix it on 0.3.x with minor changes? If no we can delay the fix for 0.4.
2. Is it still a bug on 0.4.x with new regexp? If so we really need to fix it there.

**Comment by densh** [comment]
@olafurpg Thanks for the bug report!

**Comment by LeeTibbert** [comment]
@olafurpg Thank you for the report.

@densh Yes, I will take a closer look at this.  I saw something like this over the weekend
when I was working on a regex for scalafmt.   @olafurpg and I are in correspondance
on private gitter (to reduce the noise on the main SN.)

At his request, I forked  the SN scalafmt branch and pushed rough code which might
be useful to him. More as it develops. 

The short story is that 0.3.x and 0.4.0 vary, with 0.4.0 usually being more correct.
In the sample I pushed, I tried to come up with code that world work in JVM, 0.3.x,
and 0.4.0.  

More as it develops.

**Comment by olafurpg** [comment]
If regexp has changed a lot in 0.4.x then I would not stress about fixing this issue in 0.3.x. I suspect it won't take a long time to re-publish the full dependency graph for 0.4.x once is out.

**Comment by ekrich** [comment]
I agree, I think it is more important to work on the new `re2s` in `0.4.x` to get better agreement with the JVM than worry about the compatibility between the Scala Native versions. If something is found to be wrong with `0.3.x` it probably warrants a fix if additional releases are going to be made.

**Comment by LeeTibbert** [comment]
After reviewing  my list of "promises not yet kept", I took another look at this issue from a 
rested start and the cold light of day.

#### TL:DR

I believe that this can be closed and not fix made in the 0.3.n stream.
 @olafurpg @densh, @ekrich  do you concur?

If any of you have unanswered questions and/or concerns, please let me know.

##### Evidence

*    The problem does not occur in 0.4.0-M2.  I had checked when the issue was reported
     and have just checked again.  Scalastie scala 2.13.0, 2.11.12, and SN 0.4.0-M2 all 
     give what I believe to be the intended output: newline-whitespace-verticalBar-doubleQuote.

*  The failure 0n 0.3.n appears to be in the cre2 library (ibre2-dev on Ubuntu)
   and not accessible to a simple fix.  Any fix would probably involve pre-processing strings
   sent to cre2, and that is an approach expensive to implement and prone to error.

##### Blaming the victim

SN 0.3.9 and cre2 should have done the correct replacement.  The replacement string
is passing strange; that is, more complicated than it need be.  

There are at least two independent workarounds in addition to the one posted in the 
base note.  I have tested these workarounds in my private 0.3.9 sandbox.

Agreed, code should work and devos should not need to use workarounds.

1) The first and better of the two is to remove the double backlash from the replacement string:
`.replaceAll(s"\n  |")`. The Java 8 regex documentation says that the only special characters
  in replacement strings are backslash and dollar-sign.  Vertical bar or pipe is not a special
  character and does not need quoting.

  Note also that I have removed the 's' string interpolater.  Removing the interpolater
  does nothing to solve the reported problem, but reduces complexity.

2) One can explicitly quote the replacement string.
```
val s = leadingPipeSpace.matcher("\n    |\"").replaceAll(
	Matcher.quoteReplacement(s"\n  \\|"))

```

    This is the less favored approach because 0.3.n quoteReplacement() is known, to some,
    as having quirks. To the best of my knowledge, this specific usage may not trigger
    them but other strings might.
